src/diff.rs

```diff
@@ -0,0 +1,101 @@
+use crate::undo;
+
+use std::collections::HashMap;
+use std::env;
+
+/// print a diff of the env
+pub fn run(context: bool, color: bool, shadowenv_data: &str) -> i32 {
+    let mut parts = shadowenv_data.splitn(2, ":");
+    let _prev_hash = parts.next();
+    let json_data = parts.next().unwrap_or("{}");
+    let shadowenv_data = undo::Data::from_str(json_data).unwrap();
+    let mut scalars = HashMap::new();
+    for scalar in shadowenv_data.scalars {
+        scalars.insert(scalar.name.clone(), scalar);
+    }
+    let mut lists = HashMap::new();
+    for list in shadowenv_data.lists {
+        lists.insert(list.name.clone(), list);
+    }
+    for (name, value) in env::vars() {
+        if let Some(scalar) = scalars.remove(&name) {
+            diff_scalar(&scalar, color)
+        } else if let Some(list) = lists.remove(&name) {
+            diff_list(&list, &value, color)
+        } else if context {
+            print_context(&name, &value)
+        }
+    }
+    for (_name, scalar) in &scalars {
+            diff_scalar(&scalar, color)
+    }
+    for (_name, list) in &lists {
+            diff_list(&list, &"".to_string(), color)
+    }
+    0
+}
+
+fn diff_list(list: &undo::List, current: &str, color: bool) {
```

Comment:
This seems like a lot of code. Is there a reason not to use a library like https://github.com/utkarshkukreti/diff.rs or https://github.com/johannhof/difference.rs?

Reply:
We don't actually have two texts to compare, we have lists of changes.

Reply:
Isn't it significantly easier to render the text and diff it?

Reply:
I also like the diff this lets us provide - I don't highlight the colons in path lists because they're added almost as a side effect: the path components are the more important parts, and nicely set apart this way.